lib/common/util.js

```diff
@@ -452,102 +451,228 @@ function shouldRetry(err) {
+   * @param {object|function} options - Configuration object or callback
+   *     function.
+   * @param {function=} options.onAuthorized - If provided, a request will not
+   *     be made. Instead, this function is passed the error & authorized
+   *     request options.
+   */
+  function makeAuthorizedRequest(reqOpts, callback) {
+    if (config.customEndpoint) {
+      // Using a custom API override. Do not use `google-auth-library` for
+      // authentication. (ex: connecting to a local Datastore server)
+      if (callback.onAuthorized) {
+        callback.onAuthorized(null, reqOpts);
+      } else {
+        util.makeRequest(reqOpts, callback, config);
       }
 
-      function handleRateLimitResp(err, res, body) {
-        handleResp(err, res, body, function(err, body, resp) {
-          if (shouldRetry(err) && autoRetry && MAX_RETRIES > attemptedRetries) {
-            setTimeout(function() {
-              request(authorizedReqOpts, handleRateLimitResp);
-            }, getNextRetryWait(attemptedRetries++));
-          } else {
-            callback(err, body, resp);
-          }
-        });
+      return;
+    }
+
+    util.authorizeRequest(config, reqOpts, function(err, authorizedReqOpts) {
+      if (err) {
+        (callback.onAuthorized || callback)(err);
+        return;
       }
 
       if (callback.onAuthorized) {
         callback.onAuthorized(null, authorizedReqOpts);
       } else {
-        request(authorizedReqOpts, handleRateLimitResp);
+        util.makeRequest(authorizedReqOpts, callback, config);
+      }
+    });
+  }
+
+  makeAuthorizedRequest.getCredentials = function(callback) {
+    util.getAuthClient(config, function(err, authClient) {
+      if (err) {
+        callback(err);
+        return;
       }
-    }
 
-    authorize(reqOpts, onAuthorizedRequest);
+      authClient.authorize(function(err) {
+        if (err) {
+          callback(err);
+          return;
+        }
+
+        callback(null, {
+          client_email: authClient.email,
+          private_key: authClient.key
+        });
+      });
+    });
+  };
+
+  return makeAuthorizedRequest;
+}
+
+util.makeAuthorizedRequestFactory = makeAuthorizedRequestFactory;
+
+/**
+ * Make a request through the `request` module with built-in error handling and
+ * exponential back off.
+ *
+ * @param {object} reqOpts - Request options in the format `request` expects.
+ * @param {function} callback - The callback function.
+ * @param {object=} config - Configuration object.
+ * @param {boolean=} config.autoRetry - Automatically retry requests if the
+ *     response is related to rate limits or certain intermittent server errors.
+ *     We will exponentially backoff subsequent requests by default. (default:
+ *     true)
+ * @param {number=} config.maxRetries - Maximum number of automatic retries
+ *     attempted before returning the error. (default: 3)
+ */
+function makeRequest(reqOpts, callback, config) {
```

Comment:
As stupid as it is, probably should keep callback last parameter and just do some extra checks to see if `config` is really a function.